taglocker.go

```diff
@@ -0,0 +1,212 @@
+package plc
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/dijkstracula/go-ilock"
+)
+
+// TagLocker is a plc.ReadWriter that wraps another ReadWriter, but gates
+// concurrent accesses on grabbing read or write access on a tree of locks
+// representing tag names (in the case of tree leaf nodes) and prefixes of tag
+// names (in the case of tree frond nodes).
+type TagLocker struct {
+	downstream ReadWriter
+	tagTree    *tagLockerNode
+}
+
+// ReadTag reads the given tag name from the downstream ReadWriter. If another
+// thread is concurrently writing to this tag or a prefix of the tag, we will
+// block until that thread has released its access.
+func (tl *TagLocker) ReadTag(name string, value interface{}) error {
+	components, err := ParseQualifiedTagName(name)
+	if err != nil {
+		return err
+	}
+
+	defer tl.tagTree.rUnlock(components)
+	tl.tagTree.rLock(components)
+	return tl.downstream.ReadTag(name, value)
+}
+
+// WriteTag writes the given tag value to the downstream ReadWriter. Will block
+// if another thread is reading this tag or a prefix of the tag.
+func (tl *TagLocker) WriteTag(name string, value interface{}) error {
+	components, err := ParseQualifiedTagName(name)
+	if err != nil {
+		return err
+	}
+
+	defer tl.tagTree.unlock(components)
+	tl.tagTree.lock(components)
+	return tl.downstream.WriteTag(name, value)
+}
+
+type tagLockerNode struct {
+	mtx sync.RWMutex // Ensures mutual exclusion on the fields of the node.
+
+	tagLock   *ilock.Mutex              // The logical lock that mutator threads will hold while reading and writing tags.
+	component string                    // The component of the tag name.
+	children  map[string]*tagLockerNode // All descendents of this node.
+}
+
+// NewTagLocker produces a new TagLocker.
+func NewTagLocker(downstream ReadWriter) *TagLocker {
+	var ret TagLocker
+	ret.downstream = downstream
+	ret.tagTree = newNode("/")
+	return &ret
```

Comment:
I more commonly see this as 
```go
  	return &TagLocker{
  	  	downstream: downstream,
  	  	tagTree: newNode("/"),
  	}
```

I think it's a bit clearer as it shows that the only thing you're doing is constructing and returning a single object. (I suspect it's also slightly more efficient because then the fields you're setting don't need to be set to zero, but that's unimportant.)

...of course it wouldn't work if `newNode` returned an error. Then I'd still initialize whatever I could into the struct, and then have the error-returning functions after.